src/modules/surjection/tests_impl.h

```diff
@@ -88,114 +87,112 @@ static void test_surjectionproof_api(void) {
         tmp_proof.n_inputs = 0;
-        CHECK(secp256k1_surjectionproof_generate(ctx, &tmp_proof, ephemeral_input_tags, n_inputs, &ephemeral_output_tag, 0, input_blinding_key[0], output_blinding_key) == 0);
+        CHECK(secp256k1_surjectionproof_generate(CTX, &tmp_proof, ephemeral_input_tags, n_inputs, &ephemeral_output_tag, 0, input_blinding_key[0], output_blinding_key) == 0);
     }
     CHECK(ecount == 1);
 
-    CHECK(secp256k1_surjectionproof_generate(ctx, &proof, ephemeral_input_tags, n_inputs, &ephemeral_output_tag, 0, input_blinding_key[0], output_blinding_key) == 1);
+    CHECK(secp256k1_surjectionproof_generate(CTX, &proof, ephemeral_input_tags, n_inputs, &ephemeral_output_tag, 0, input_blinding_key[0], output_blinding_key) == 1);
 
     /* Check serialize */
     ecount = 0;
     serialized_len = sizeof(serialized_proof);
-    CHECK(secp256k1_surjectionproof_serialize(ctx, serialized_proof, &serialized_len, &proof) != 0);
+    CHECK(secp256k1_surjectionproof_serialize(CTX, serialized_proof, &serialized_len, &proof) != 0);
     CHECK(ecount == 0);
     serialized_len = sizeof(serialized_proof);
-    CHECK(secp256k1_surjectionproof_serialize(ctx, NULL, &serialized_len, &proof) == 0);
+    CHECK(secp256k1_surjectionproof_serialize(CTX, NULL, &serialized_len, &proof) == 0);
     CHECK(ecount == 1);
     serialized_len = sizeof(serialized_proof);
-    CHECK(secp256k1_surjectionproof_serialize(ctx, serialized_proof, NULL, &proof) == 0);
+    CHECK(secp256k1_surjectionproof_serialize(CTX, serialized_proof, NULL, &proof) == 0);
     CHECK(ecount == 2);
     serialized_len = sizeof(serialized_proof);
-    CHECK(secp256k1_surjectionproof_serialize(ctx, serialized_proof, &serialized_len, NULL) == 0);
+    CHECK(secp256k1_surjectionproof_serialize(CTX, serialized_proof, &serialized_len, NULL) == 0);
     CHECK(ecount == 3);
 
     serialized_len = sizeof(serialized_proof);
-    CHECK(secp256k1_surjectionproof_serialize(ctx, serialized_proof, &serialized_len, &proof) != 0);
+    CHECK(secp256k1_surjectionproof_serialize(CTX, serialized_proof, &serialized_len, &proof) != 0);
     /* Check parse */
-    CHECK(secp256k1_surjectionproof_parse(ctx, &proof, serialized_proof, serialized_len) != 0);
+    CHECK(secp256k1_surjectionproof_parse(CTX, &proof, serialized_proof, serialized_len) != 0);
     CHECK(ecount == 3);
-    CHECK(secp256k1_surjectionproof_parse(ctx, NULL, serialized_proof, serialized_len) == 0);
+    CHECK(secp256k1_surjectionproof_parse(CTX, NULL, serialized_proof, serialized_len) == 0);
     CHECK(ecount == 4);
-    CHECK(secp256k1_surjectionproof_parse(ctx, &proof, NULL, serialized_len) == 0);
+    CHECK(secp256k1_surjectionproof_parse(CTX, &proof, NULL, serialized_len) == 0);
     CHECK(ecount == 5);
-    CHECK(secp256k1_surjectionproof_parse(ctx, &proof, serialized_proof, 0) == 0);
+    CHECK(secp256k1_surjectionproof_parse(CTX, &proof, serialized_proof, 0) == 0);
     CHECK(ecount == 5);
-
-    secp256k1_context_destroy(sttc);
```

Comment:
This should reset all the callbacks in the contexts (similar to `run_ec_illegal_argument_tests` in tests.c):

```
    secp256k1_context_set_error_callback(CTX, NULL, NULL);
    secp256k1_context_set_error_callback(STATIC_CTX, NULL, NULL);
    secp256k1_context_set_illegal_callback(CTX, NULL, NULL);
    secp256k1_context_set_illegal_callback(STATIC_CTX, NULL, NULL);
```

Reply:
I intentionally leftt this out because it's difficult to enforce and useless, but then I remembered https://github.com/bitcoin-core/secp256k1/issues/1167.

Reply:
Weirdly, in the upstream code we often only reset STATIC_CTX, but not CTX (e.g. in schnorrsig and extrakeys).

Reply:
All the callbacks should be correctly reset now.

Reply:
> Weirdly, in the upstream code we often only reset STATIC_CTX, but not CTX (e.g. in schnorrsig and extrakeys).

Hm, I don't think there's a deep reason for this. I assume these are simply code smells in the upstream tests. A solution of bitcoin-core/secp256k1#1167 will probably fix this on the way.